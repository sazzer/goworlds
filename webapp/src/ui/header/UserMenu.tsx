/** The props that a HeaderBar needs */
import React, {FunctionComponent} from "react";
import {useTranslation} from "react-i18next";
import {Dropdown, Icon} from "semantic-ui-react";
import {shallowEqual, useDispatch, useSelector} from "react-redux";
import {selectCurrentUserId} from "../../authentication/currentUserId";
import {loadUser, selectUserById} from "../../users/users";
import {Link} from "react-router-dom";
import {logout} from "../../authentication/logout";
import {User} from "../../users/user";

type UserMenuProps = {};

/**
 * The User menu in the application
 * @constructor
 */
export const UserMenu: FunctionComponent<UserMenuProps> = () => {
    const { t } = useTranslation();
    const dispatch = useDispatch();
    const currentUserId = useSelector(selectCurrentUserId);
    const currentUser : User | undefined = useSelector(selectUserById(currentUserId), shallowEqual);

    if (currentUser === undefined) {
        dispatch(loadUser(currentUserId));
    }

    return (
        <Dropdown item text={currentUser === undefined ? '' : currentUser.name} loading={currentUser === undefined} data-test="UserMenu">
            <Dropdown.Menu>
                <Link to="/profile" className="ui item">
                    <Icon name="edit" />
                    {t('userMenu.editProfile')}
                </Link>
                <Dropdown.Divider />
                <div className="ui item" onClick={() => dispatch(logout())}>
                    <Icon name="log out" />
                    {t('userMenu.logOut')}
                </div>
            </Dropdown.Menu>
        </Dropdown>
    );
};
