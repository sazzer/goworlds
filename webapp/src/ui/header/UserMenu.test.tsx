import React from 'react';
import {render, fireEvent} from '@testing-library/react';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {UserMenu} from './UserMenu';
import {MemoryRouter} from "react-router";

function setup(state: any) {
    const mockStoreCreator = configureStore();
    const store = mockStoreCreator(state);

    const element = render(<Provider store={store}><MemoryRouter><UserMenu /></MemoryRouter></Provider>);

    const logout = () => new Promise(resolve => {
        fireEvent.click(element.getByText('Log Out'));
        setImmediate(resolve);
    });

    return {
        store,

        logout,

        element,
    }
}

it('Renders correctly when a user ID is available but not yet loaded', () => {
    const {element, store} = setup({
        currentUserId: {
            userId: 'abc123',
        },
        users: {
            users: []
        }
    });

    expect(element.baseElement).toMatchSnapshot();
    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()).toContainEqual(
        {
            'type': 'Users/loadUser',
            'payload': {
                'userId': 'abc123',
                'force': false,
            },
        }
    )
});

it('Renders correctly when a user ID is available and loaded', () => {
    const {element, store} = setup({
        currentUserId: {
            userId: 'abc123',
        },
        users: {
            users: [
                {
                    id: 'abc123',
                    name: 'Graham',
                }
            ]
        }
    });

    expect(element.baseElement).toMatchSnapshot();
    expect(store.getActions()).toHaveLength(0);
});

it('Dispatches the correct action when logging out', () => {
    const {logout, store} = setup({
        currentUserId: {
            userId: 'abc123',
        },
        users: {
            users: [
                {
                    id: 'abc123',
                    name: 'Graham',
                }
            ]
        }
    });

    logout();
    expect(store.getActions()).toHaveLength(1);
    expect(store.getActions()).toContainEqual({
        type: 'Authentication/logout',
        payload: {}
    });
});
