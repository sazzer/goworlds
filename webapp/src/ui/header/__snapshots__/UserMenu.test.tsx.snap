// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly when a user ID is available and loaded 1`] = `
<body>
  <div>
    <div
      aria-busy="false"
      aria-expanded="false"
      class="ui item dropdown"
      data-test="UserMenu"
      role="listbox"
      tabindex="0"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        class="text"
        role="alert"
      >
        Graham
      </div>
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <a
          class="ui item"
          href="/profile"
        >
          <i
            aria-hidden="true"
            class="edit icon"
          />
          Edit Profile
        </a>
        <div
          class="divider"
        />
        <div
          class="ui item"
        >
          <i
            aria-hidden="true"
            class="log out icon"
          />
          Log Out
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Renders correctly when a user ID is available but not yet loaded 1`] = `
<body>
  <div>
    <div
      aria-busy="true"
      aria-expanded="false"
      class="ui loading item dropdown"
      data-test="UserMenu"
      role="listbox"
      tabindex="0"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        class="text"
        role="alert"
      />
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <a
          class="ui item"
          href="/profile"
        >
          <i
            aria-hidden="true"
            class="edit icon"
          />
          Edit Profile
        </a>
        <div
          class="divider"
        />
        <div
          class="ui item"
        >
          <i
            aria-hidden="true"
            class="log out icon"
          />
          Log Out
        </div>
      </div>
    </div>
  </div>
</body>
`;
